language/move-prover/src/prelude.bpl

```diff
@@ -517,27 +518,42 @@ function {:inline} $InRange(r: $Value, i: int): bool {
 // Memory
 
 type {:datatype} $Location;
-function {:constructor} $Global(t: $TypeValue, a: int): $Location;
+
+// A global resource location within the statically known resource type's memory.
+// `ts` are the type parameters for the outer type, and `a` is the address.
+function {:constructor} $Global(ts: $TypeValueArray, a: int): $Location;
+
+// A local location. `i` is the unique index of the local.
 function {:constructor} $Local(i: int): $Location;
+
+// The location of a reference outside of the verification scope, for example, a `&mut` parameter
+// of the function being verified. References with these locations don't need to be written back
+// when mutation ends.
 function {:constructor} $Param(i: int): $Location;
 
-type {:datatype} $Reference;
-function {:constructor} $Reference(l: $Location, p: $Path, v: $Value): $Reference;
-const $DefaultReference: $Reference;
 
+// A mutable reference which also carries its current value. Since mutual references
+// are single threaded in Move, we can keep them together and treat them as a value
+// during mutation until the point they are stored back to their original location.
+type {:datatype} $Mutation;
+function {:constructor} $Mutation(l: $Location, p: $Path, v: $Value): $Mutation;
+const $DefaultMutation: $Mutation;
+
+// Representation of memory for a given type. The maps take the content of a Global location.
 type {:datatype} $Memory;
-function {:constructor} $Memory(domain: [$Location]bool, contents: [$Location]$Value): $Memory;
+function {:constructor} $Memory(domain: [$TypeValueArray, int]bool, contents: [$TypeValueArray, int]$Value): $Memory;
```

Comment:
We should remember to ask SMT experts whether there is a cost to multi-dimensional arrays.

Reply:
... where we should add: compared to a single variant datatype with two fields (tuple).